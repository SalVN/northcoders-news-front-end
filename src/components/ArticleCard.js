import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

const ArticleCard = function (props) {
  return (
    <div className='box'>
      <article className='media'>
        <div className='media-left'>
          {props.number + 1}
        </div>
        <div className='media-left'>
          <button className='button' onClick={props.voteHandler.bind(this, 'up', props.id)}><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
          <div>{props.votes}</div>
          <button className='button' onClick={props.voteHandler.bind(this, 'down', props.id)}><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
        </div>
        <div className='media-content'>
          <div className='content'>
            <Link to={`/articles/${props.id}`}>
              <h3 className='title is-3'>{props.title}</h3>
            </Link>
            <p>
              <span>added by: {props.author}</span>
              <span>tags: {props.tags}</span>
              <span>comments: {props.comment_count}</span>
            </p>
          </div>
        </div>
      </article>
    </div>
  );
};

ArticleCard.propTypes = {
  title: PropTypes.string.isRequired,
  votes: PropTypes.number.isRequired,
  number: PropTypes.number.isRequired,
  author: PropTypes.string.isRequired,
  id: PropTypes.string.isRequired,
  tags: PropTypes.string.isRequired,
  comment_count: PropTypes.number.isRequired,
  voteHandler: PropTypes.func.isRequired
};


export default ArticleCard;
