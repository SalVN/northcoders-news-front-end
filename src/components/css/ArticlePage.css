.article-page-hr {
    border: 1px solid #c1002e;
    padding: 0px, 30px;
}