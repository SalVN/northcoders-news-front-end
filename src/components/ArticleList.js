import React from 'react';
import ArticleCard from './ArticleCard';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import * as actions from '../actions/actions';

class ArticleList extends React.Component {
  componentDidMount() {
    if (this.props.users && this.props.users.length < 1) {
      this.props.fetchUsers();
    }
  }
  render() {
    const users = this.props.users;
    const voteHandler = this.props.voteArticle;
    return (
      <div id='ArticleList' className='content'>
        {this.props.articles.length > 0 &&
          this.props.articles.map((article, i) => {
            const username = article.created_by;
            let index;
            users &&
              users.forEach((user, i) => {
                if (user.username === username) index = i;
              });
            return (
              <ArticleCard
                key={article._id}
                title={article.title}
                number={i}
                votes={article.votes}
                author={article.created_by}
                id={article._id}
                tags={article.belongs_to}
                comment_count={article.comment_count}
                voteHandler={voteHandler}
                userData={users[index]}
              />
            );
          })
        }
      </div>
    );
  }
}

function mapDispatchToProps(dispatch) {
  return {
    fetchUsers: () => {
      dispatch(actions.fetchUsers());
    }
  };
}

function mapStateToProps(state) {
  return {
    loading: state.users.loading,
    users: state.users.users
  };
}

ArticleList.propTypes = {
  articles: PropTypes.array.isRequired,
  loading: PropTypes.bool.isRequired,
  fetchUsers: PropTypes.func.isRequired,
  voteArticle: PropTypes.func.isRequired,
  users: PropTypes.array.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(ArticleList);