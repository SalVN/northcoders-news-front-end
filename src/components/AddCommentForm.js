import React from 'react';
import PropTypes from 'prop-types';

import './css/AddCommentForm.css';

class AddCommentForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            text: ''
        };
        this.handleChange = this.handleChange.bind(this);
    }
    render() {
        return (
            <div>
                {this.props.showForm
                    ?
                    <div className='box box-style box-style-comment box-style-comment'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <form onSubmit={this.props.handleSubmit}>
                                        <label className='comment-form-heading' htmlFor="comment-input">Join the conversation:</label>
                                        <br />
                                        <div className='comment-form' onBlur={this.props.toggleForm.bind(this, this.state.text)}>
                                            <textarea className='add-comment-textarea' onChange={this.handleChange} id="comment-input" defaultValue={this.state.text} placeholder='Join the conversation' />
                                        </div>
                                        <div className='button-div'>
                                            <button className='button add-comment-submit' type='submit'>Submit</button>
                                        </div>
                                    </form>
                                </div>
                            </div>
                        </article>
                    </div>
                    : <div className='box box-style box-style-comment box-style-comment-small'>
                        <article className='media'>
                            <div className='media-left'>
                                <img className='comment-form-image' src={this.props.user.avatar_url} />
                            </div>
                            <div className='media-content'>
                                <div className='content'>
                                    <div onClick={this.props.toggleForm.bind(this, null)} className='comment-form-small'>
                                        <p>Join the conversation</p>
                                    </div>
                                </div>
                            </div>
                        </article>
                    </div>
                }
            </div>
        );
    }
    handleChange(e) {
        this.setState({
            text: e.target.value
        });
    }
}

AddCommentForm.propTypes = {
    id: PropTypes.string.isRequired,
    addComment: PropTypes.func.isRequired,
    toggleForm: PropTypes.func.isRequired,
    user: PropTypes.object.isRequired,
    showForm: PropTypes.bool.isRequired,
    handleSubmit: PropTypes.func.isRequired
};

export default AddCommentForm;