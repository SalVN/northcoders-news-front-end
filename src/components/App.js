import React, {Component} from 'react';
import PropTypes from 'prop-types';

import TopicNavBar from './TopicNavBar';
import MainBanner from './MainBanner';
import UserCard from './UserCard';
import UserListCard from './UserListCard';

import './css/App.css';

export class App extends Component {
  render() {
    let pageStyling = window.innerWidth < 800 ? 'main-page-mobile' : 'main-page';
    return (
      <div className='page'>
        <TopicNavBar />
        <MainBanner />
        <div className={pageStyling}>
          <div className={'columns is-gapless'}>
            <div className='column is-three-quarters'>
              {this.props.children}
            </div>
            <div className='column is-one-quarter'>
              <div className='is-hidden-mobile'>
                <UserCard />
              </div>
              <UserListCard />
            </div>
          </div>
        </div>
      </div>
    );
  }
}

App.propTypes = {
  children: PropTypes.element.isRequired
};

export default App;
