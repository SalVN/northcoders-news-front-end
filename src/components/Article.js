import React from 'react';
import PropTypes from 'prop-types';
import { BrowserRouter as Router, Link } from 'react-router-dom';

import './css/Article.css';

const Article = function (props) {
    return (
        <div className='box'>
            <div className='columns'>
                <div className='column is-1'>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-up" aria-hidden="true"></i></button>
                    <p className='comment-card-votes'><strong>{props.article.votes}</strong></p>
                    <button className='button is-black is-inverted is-small'><i className="fa fa-arrow-down" aria-hidden="true"></i></button>
                </div>
                <div className='column'>
                    <div className='title is-3 article-title'>{props.article.title}</div>
                    <p>
                        <span>
                            <span className='article-subheading-key'>added by:</span>
                            <Link to='#'>
                                <span>
                                    {props.user &&
                                        <img className='article-card-avatar' src={props.user.avatar_url} alt='user avatar' />
                                    }
                                </span>
                                <span className='article-subheading-value'>
                                    <strong>{props.article.created_by}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>tags:</span>
                            <Link to={`/topics/${props.article.belongs_to}/articles`}>
                                <span className='article-card-subheading-value'>
                                    <strong>{props.article.belongs_to}</strong>
                                </span>
                            </Link>
                        </span>
                        <span>
                            <span className='article-card-subheading-key'>comments:</span>
                            {/*<Link to='#'>*/}
                            <span className='article-card-subheading-value-notlink'>
                                <strong>{props.article.commentCount}</strong>
                            </span>
                            {/*</Link>*/}
                        </span>
                    </p>
                    <div>{props.article.body}</div>
                </div>
            </div>
        </div>
    );
};

Article.propTypes = {
    article: PropTypes.object.isRequired,
    user: PropTypes.object.isRequired
};

export default Article;