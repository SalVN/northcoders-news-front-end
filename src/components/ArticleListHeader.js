import React from 'react';
import PropTypes from 'prop-types';

import './css/MainArticleList.css';

const ArticleListHeader = function (props) {
    return (
        <div className='content'>
            <div className='columns'>
                <div className='column is-3'>
                    <h2 className='page-title'><strong>{props.title}</strong></h2>
                </div>
                <div className='column is-1 is-offset-5 label'>
                    <label htmlFor="Sort by">Sort by:</label>
                </div>
                <div className='column is-3'>
                    <div className='select is-primary select-menu'>
                        <select name="Sort by" onChange={props.handleClickSelect}>
                            <option value="votes">Popularity</option>
                            <option value="comments">Number of Comments</option>
                            <option value="random">Random</option>
                        </select>
                    </div>
                </div>
            </div>
        </div>
    );
};

ArticleListHeader.propTypes = {
    title: PropTypes.string.isRequired,
    handleClickSelect: PropTypes.func.isRequired
};

export default ArticleListHeader;