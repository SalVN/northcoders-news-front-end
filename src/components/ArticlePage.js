import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { BrowserRouter as Router, Redirect } from 'react-router-dom';

import * as actions from '../actions/actions';
import { getIndex } from '../utilities/getIndex';

import Article from './Article';
import Comments from './Comments';
import './css/ArticlePage.css';

class ArticlePage extends Component {
    constructor(props) {
        super(props);
        this.articleVoteHandler = this.articleVoteHandler.bind(this);
    }
    componentDidMount() {
        if (this.props.articles && this.props.articles.length < 1) {
            this.props.fetchArticles();
        }
        if (this.props.users && this.props.users.length < 1) {
            this.props.fetchUsers();
        }
    }
    render() {
        let article;
        let user;
        let articleNotFound = false;
        if (this.props.articles) {
            article = this.props.articles.reduce((acc, article) => {
                if (article._id === this.props.match.params.id) acc = article;
                return acc;
            }, {});
            if (this.props.articles.length > 0 && article && !article._id) {
                articleNotFound = true;
            }
        }
        if (this.props.users && article) {
            user = this.props.users.reduce((acc, user) => {
                if (article.created_by === user.username) acc = user;
                return acc;
            }, {});
        }
        const index = getIndex(this.props.articles, article._id);
        return (
            <div className='article-page'>
                {articleNotFound &&
                    <Redirect to='/not-found' />
                }
                {
                    this.props.articlesLoading
                        ? <span>
                            <i className='fa fa-refresh fa-spin' />
                        </span>
                        : <Article
                            article={article}
                            user={user}
                            voteArticle={this.articleVoteHandler}
                            index={index}
                            articles={this.props.articles}
                        />
                }
                <hr className='article-page-hr' />
                <Comments
                    id={this.props.match.params.id}
                    users={this.props.users}
                />
            </div>
        );
    }
    articleVoteHandler(vote, id) {
        this.props.voteArticle(vote, id);
    }
}

function mapDispatchToProps(dispatch) {
    return {
        fetchArticles: () => {
            dispatch(actions.fetchArticles());
        },
        fetchUsers: () => {
            dispatch(actions.fetchUsers());
        },
        voteArticle: (vote, id) => {
            dispatch(actions.voteArticle(vote, id));
        }
    };
}

function mapStateToProps(state) {
    return {
        articles: state.articles.articles,
        articlesLoading: state.articles.loading,
        users: state.users.users
    };
}

ArticlePage.propTypes = {
    articles: PropTypes.array.isRequired,
    fetchArticles: PropTypes.func.isRequired,
    match: PropTypes.object.isRequired,
    users: PropTypes.array.isRequired,
    fetchUsers: PropTypes.func.isRequired,
    voteArticle: PropTypes.func.isRequired,
    articlesLoading: PropTypes.bool.isRequired
};

export default connect(mapStateToProps, mapDispatchToProps)(ArticlePage);