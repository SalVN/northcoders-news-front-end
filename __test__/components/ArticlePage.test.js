import React from 'react';
import { shallow } from 'enzyme';
import { Provider } from 'react-redux';
import configureStore from 'redux-mock-store';
import { MemoryRouter } from 'react-router-dom';
import ReactShallowRenderer from 'react-test-renderer/shallow';

const mockStore = configureStore();
const initialState = {
    comments: {
        comments: [],
        loading: false
    },
    oneUser: {
        oneUser: {
            _id: '59b1b18b327cce1fb043bdb1',
            username: 'northcoder',
            name: 'Awesome Northcoder',
            avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
            comments_vote_count: 8,
            articles_vote_count: 5,
            __v: 0
        },
        loading: false
    }
};

import { ArticlePage } from '../../src/components/ArticlePage';

describe('ArticlePage', () => {
    const articles = [{
        _id: '59b01acf006c8dbca914672f',
        title: 'Football is fun',
        body: 'something',
        belongs_to: 'football',
        __v: 0,
        votes: 3,
        comment_count: 0
    }];
    const users = [{
        _id: '59b1b18b327cce1fb043bdb2',
        username: 'northcoder',
        name: 'Awesome Northcoder',
        avatar_url: 'https://avatars3.githubusercontent.com/u/6791502?v=3&s=200',
        __v: 0
    }];
    it('is a function', () => {
        expect(typeof ArticlePage).toEqual('function');
    });

    it('renders', () => {
        const store = mockStore(initialState);
        const wrapper = shallow(<ArticlePage
            store={store}
            articles={articles}
            fetchArticles={x => x}
            match={
                {
                    params: {
                        id: '59b01acf006c8dbca914672f'
                    }
                }}
            users={users}
            fetchUsers={x => x}
            voteArticle={x => x}
            articlesLoading={false}
        />);
        expect(wrapper.children().length).toEqual(3);
    });

    it('renders correctly', () => {
        const store = mockStore(initialState);
        const renderer = new ReactShallowRenderer();
        const tree = renderer.render(
            <Provider store={store}>
                <MemoryRouter>
                    <ArticlePage
                        store={store}
                        articles={articles}
                        fetchArticles={x => x}
                        match={
                            {
                                params: {
                                    id: '59b01acf006c8dbca914672f'
                                }
                            }}
                        users={users}
                        fetchUsers={x => x}
                        voteArticle={x => x}
                        articlesLoading={false} />
                </MemoryRouter>
            </Provider>
        );
        expect(tree).toMatchSnapshot();
    });

    it('redirects if the article is not found', () => {
        const store = mockStore(initialState);
        const wrapperA = shallow(<ArticlePage
            store={store}
            articles={articles}
            fetchArticles={x => x}
            match={
                {
                    params: {
                        id: '59b01acf006c8dbca914672f'
                    }
                }}
            users={users}
            fetchUsers={x => x}
            voteArticle={x => x}
            articlesLoading={false}
        />);
        expect(wrapperA.find('Redirect').length).toEqual(0);

        const wrapperB = shallow(<ArticlePage
            store={store}
            articles={articles}
            fetchArticles={x => x}
            match={
                {
                    params: {
                        id: 'abc'
                    }
                }}
            users={users}
            fetchUsers={x => x}
            voteArticle={x => x}
            articlesLoading={false}
        />);
        expect(wrapperB.find('Redirect').length).toEqual(1);
    });

    it('renders a loading icon if this.props.articlesLoading', () => {
        const store = mockStore(initialState);
        const enzymeWrapperA = shallow(<ArticlePage
            store={store}
            articles={articles}
            fetchArticles={x => x}
            match={
                {
                    params: {
                        id: 'abc'
                    }
                }}
            users={users}
            fetchUsers={x => x}
            voteArticle={x => x}
            articlesLoading={false}
        />);
        expect(enzymeWrapperA.find('.fa-spin').length).toBe(0);

        const enzymeWrapperB = shallow(<ArticlePage
            store={store}
            articles={articles}
            fetchArticles={x => x}
            match={
                {
                    params: {
                        id: 'abc'
                    }
                }}
            users={users}
            fetchUsers={x => x}
            voteArticle={x => x}
            articlesLoading={true}
        />);
        expect(enzymeWrapperB.find('.fa-spin').length).toBe(1);
    });

    it('only renders the article if !this.props.articleLoading', () => {
        const store = mockStore(initialState);
        const enzymeWrapperA = shallow(<ArticlePage
            store={store}
            articles={articles}
            fetchArticles={x => x}
            match={
                {
                    params: {
                        id: '59b01acf006c8dbca914672f'
                    }
                }}
            users={users}
            fetchUsers={x => x}
            voteArticle={x => x}
            articlesLoading={false}
        />);
        expect(enzymeWrapperA.find('Article').length).toBe(1);

        const enzymeWrapperB = shallow(<ArticlePage
            store={store}
            articles={articles}
            fetchArticles={x => x}
            match={
                {
                    params: {
                        id: '59b01acf006c8dbca914672f'
                    }
                }}
            users={users}
            fetchUsers={x => x}
            voteArticle={x => x}
            articlesLoading={true}
        />);
        expect(enzymeWrapperB.find('Article').length).toBe(0);
    });
});