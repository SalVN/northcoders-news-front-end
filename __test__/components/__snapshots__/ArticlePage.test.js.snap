// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticlePage renders correctly 1`] = `
<MemoryRouter>
  <ArticlePage
    articles={
      Array [
        Object {
          "__v": 0,
          "_id": "59b01acf006c8dbca914672f",
          "belongs_to": "football",
          "body": "something",
          "comment_count": 0,
          "title": "Football is fun",
          "votes": 3,
        },
      ]
    }
    articlesLoading={false}
    fetchArticles={[Function]}
    fetchUsers={[Function]}
    match={
      Object {
        "params": Object {
          "id": "59b01acf006c8dbca914672f",
        },
      }
    }
    store={
      Object {
        "clearActions": [Function],
        "dispatch": [Function],
        "getActions": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
      }
    }
    users={
      Array [
        Object {
          "__v": 0,
          "_id": "59b1b18b327cce1fb043bdb2",
          "avatar_url": "https://avatars3.githubusercontent.com/u/6791502?v=3&s=200",
          "name": "Awesome Northcoder",
          "username": "northcoder",
        },
      ]
    }
    voteArticle={[Function]}
  />
</MemoryRouter>
`;
